### PR TITLE
Added default connect_timeout  in cluster config

### LINKED ISSUE #16439: Required field not comment in docs
*Title*:  Required field not comment in docs

*Description*:
>What issue is being seen? Describe what should be happening instead of
the bug, for example: Envoy should not crash, the expected value isn't
returned, etc.

I find if no connection_timeout set, cluster config would raise an error.
In https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#, it does not comment `REQUIRED`.

Some [discuss](https://envoyproxy.slack.com/archives/C78HA81DH/p1620746018086800?thread_ts=1620716123.084300&cid=C78HA81DH) is in slack.

I want to fix it while do not know how to do :(

*Repro steps*:
> Include sample requests, environment, etc. All data and inputs
required to reproduce the bug.

>**Note**: The [Envoy_collect tool](https://github.com/envoyproxy/envoy/blob/main/tools/envoy_collect/README.md)
gathers a tarball with debug logs, config and the following admin
endpoints: /stats, /clusters and /server_info. Please note if there are
privacy concerns, sanitize the data prior to sharing the tarball/pasting.

*Admin and Stats Output*:
>Include the admin output for the following endpoints: /stats,
/clusters, /routes, /server_info. For more information, refer to the
[admin endpoint documentation.](https://www.envoyproxy.io/docs/envoy/latest/operations/admin)

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Config*:
>Include the config used to configure Envoy.

*Logs*:
>Include the access logs and the Envoy logs.

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Call Stack*:
> If the Envoy binary is crashing, a call stack is **required**.
Please refer to the [Bazel Stack trace documentation](https://github.com/envoyproxy/envoy/tree/main/bazel#stack-trace-symbol-resolution).

### LINKED ISSUE #16583: proto validation: Cluster message should require "connect_timeout" field be set
Cluster message should require "connect_timeout" instead of only checking the duration is greater-than 0.

## Description
*I'm not sure if this belongs here or in https://github.com/envoyproxy/protoc-gen-validate. Feel free to redirect me there if it's more appropriate*

The `Cluster` message in cluster.proto does not require the `connect_timeout` field as part of the proto validation. It only requires it to be greater than the default duration (which I assume is 0):
https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/cluster/v3/cluster.proto#L712

When dealing with Go-compiled Cluster messages that don't have the `connect_timeout` field set, we can call `Validate()` on them without error. This is because the generated code only cares so validate if the field has been set. It doesn't consider the default value since the duration is nil:
```
	if d := m.GetConnectTimeout(); d != nil {
		dur, err := ptypes.Duration(d)
		if err != nil {
			return ClusterValidationError{
				field:  "ConnectTimeout",
				reason: "value is not a valid duration",
				cause:  err,
			}
		}

		gt := time.Duration(0*time.Second + 0*time.Nanosecond)

		if dur <= gt {
			return ClusterValidationError{
				field:  "ConnectTimeout",
				reason: "value must be greater than 0s",
			}
		}

	}
```

This would be trivial to fix and I'm more than happy to just fire off the PR. I'm just not clear on the API implications and whether it would break anything.

## Repro steps
I'm running a fairly recent main build:
`./envoy_static  version: 2026543d30272e816bd1b46d1e99d50b2212d80f/1.19.0-dev/Modified/DEBUG/BoringSSL`

I can successfully validate the following static configuration with that binary:
```
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
    - filters:

  clusters:
  - name: testing123
    type: LOGICAL_DNS
    connect_timeout: 1s
    load_assignment:
      cluster_name: service_envoyproxy_io
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.envoyproxy.io
                port_value: 443
```

With the following output:
```
[2021-05-19 21:13:03.586][124567][info][main] [source/server/server.cc:667] runtime: layers:
  - name: base
    static_layer:
      {}
  - name: admin
    admin_layer:
      {}
[2021-05-19 21:13:03.586][124567][info][config] [source/server/configuration_impl.cc:128] loading tracing configuration
[2021-05-19 21:13:03.586][124567][info][config] [source/server/configuration_impl.cc:88] loading 0 static secret(s)
[2021-05-19 21:13:03.586][124567][info][config] [source/server/configuration_impl.cc:94] loading 1 cluster(s)
[2021-05-19 21:13:03.592][124567][info][config] [source/server/configuration_impl.cc:98] loading 1 listener(s)
[2021-05-19 21:13:03.594][124567][info][config] [source/server/configuration_impl.cc:110] loading stats configuration
configuration 'good.yaml' OK
```

However, removing the `connect_timeout` field results in a failure to validate.
```
[txallen@dev-dsk-txallen-2a-95bb9472 ~]$ diff -u3 good.yaml bad.yaml
--- good.yaml	2021-05-19 21:13:00.416024933 +0000
+++ bad.yaml	2021-05-19 21:13:19.331503991 +0000
@@ -11,7 +11,6 @@
   clusters:
   - name: testing123
     type: LOGICAL_DNS
-    connect_timeout: 1s
     load_assignment:
       cluster_name: service_envoyproxy_io
       endpoints:
```
```
[txallen@dev-dsk-txallen-2a-95bb9472 ~]$ ./envoy_static --mode validate -c bad.yaml 
[2021-05-19 21:21:03.242][128680][info][main] [source/server/server.cc:667] runtime: layers:
  - name: base
    static_layer:
      {}
  - name: admin
    admin_layer:
      {}
[2021-05-19 21:21:03.243][128680][info][config] [source/server/configuration_impl.cc:128] loading tracing configuration
[2021-05-19 21:21:03.243][128680][info][config] [source/server/configuration_impl.cc:88] loading 0 static secret(s)
[2021-05-19 21:21:03.243][128680][info][config] [source/server/configuration_impl.cc:94] loading 1 cluster(s)
[2021-05-19 21:21:03.246][128680][critical][main] [source/server/config_validation/server.cc:62] error initializing configuration 'bad.yaml': Field 'connect_timeout' is missing in: name: "testing123"
type: LOGICAL_DNS
load_assignment {
  cluster_name: "service_envoyproxy_io"
  endpoints {
    lb_endpoints {
      endpoint {
        address {
          socket_address {
            address: "www.envoyproxy.io"
            port_value: 443
          }
        }
      }
    }
  }
}
183412668: "envoy.api.v2.Cluster"
```

### PR DESCRIPTION
**Pull request by mk46**
Signed-off-by: Manish Kumar <manish.kumar1@india.nec.com>

Added default value of 5s for  connect_timeout in the cluster config
 
 Fixes #16439 

Fixes #16583

### CODE REVIEW

**Comment by phlax** [comment]
@mk46 looks good, but i would like to check that it really is required

afaict from quick grep all of the config has this setting - so it seems like it is required as @daixiang0 has found - i would like to confirm this 100%

the other thing im wondering is if it isnt marked required already in the api - is the problem  with the (api) docs or should it have some default in the actual code 

cc @htuch

**Comment by phlax** [comment]
@mk46 not sure why exactly - but looks like CI doesnt like this change

**Comment by mk46** [comment]
I had fixed the CI error in `CdsApiImplTest`  but not sure about `ExampleConfigsTest`.

**Comment by phlax** [comment]
not 100% sure but looking here https://dev.azure.com/cncf/envoy/_build/results?buildId=75433&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=6b0ace90-28b2-51d9-2951-b1fd35e67dec&l=21656

...it looks to me like making this required is also enforcing for xDS (eds) and not just bootstrap config - afaict whatever the situation with bootstrap it looks as though its not currently required for non-boostrap config of clusters

**Comment by mk46** [comment]
@phlax @htuch  Please take a look.

**Comment by lizan** [comment]
Can this be tested somehow?

**Comment by mk46** [comment]
Hi @lizan, I have added a unit test for checking the default value.  Please take a look.

**Comment by mk46** [comment]
I am not sure why GCC is failing, which is not related to current changes.

**Comment by phlax** [comment]
> I am not sure why GCC is failing, which is not related to current changes.

its a problem on main - its being worked on here #16484

**Review by htuch — APPROVED** [review]
LGTM

**Review by wrowe — DISMISSED** [review]
This change is solid, but missing docs of the new default. Perhaps just a tracking issue to ensure the docs are refreshed if this is merged first?

https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto.html
https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/timeouts

**Review by htuch — APPROVED** [review]

File: api/envoy/config/cluster/v3/cluster.proto [hunk 1]
```diff
@@ -709,6 +709,7 @@ message Cluster {
   EdsClusterConfig eds_cluster_config = 3;
 
   // The timeout for new network connections to hosts in the cluster.
+  // If not set, a default value of 5s will be used.
```
Comment (htuch) [inline api/envoy/config/cluster/v3/cluster.proto:712]: @mk46 actually, can you make sure @wrowe concerns are met? I think https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto.html is handled automatically by this, but there was the ask on FAQ.

Also, maybe add a version history entry. Thanks.

Comment (mk46) [reply api/envoy/config/cluster/v3/cluster.proto:712]: added in version history and FAQ.

**Comment by phlax** [comment]
docs rendered here https://storage.googleapis.com/envoy-pr/16453/docs/index.html

**Pull request merged**